src/lib.rs

```diff
@@ -1362,4 +1445,82 @@ mod tests {
             .to_value();
         assert_eq!(Value::Number(3.0), value);
     }
+
+    #[cfg_attr(feature = "derive", duktape_fn)]
+    fn test_rust_fn(input: u8) -> String {
+        format!("test {}", input)
+    }
+
+    #[cfg_attr(feature = "derive", duktape_fn)]
+    fn test_rust_complex_fn(input: TestComplexStruct) -> TestComplexStruct {
+        println!("{:?}", input);
+        input
+    }
+
+    #[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+    enum TestEnum {
+        A,
+    }
+
+    #[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+    enum TestComplexEnum {
+        A(i64, i64),
+        B { hello: String },
+    }
+
+    #[derive(Debug, serde::Deserialize, serde::Serialize, PartialEq)]
+    struct TestComplexStruct {
+        a: i64,
+        b: String,
+        c: bool,
+        d: TestEnum,
+        e: TestComplexEnum,
+        f: TestComplexEnum,
+        g: std::collections::HashMap<String, String>,
+    }
+
+    #[test]
+    fn call_rs_from_js() {
+        let ctx = Context::new();
+
+        add_global_fn!(ctx, test_rust_fn);
```

Comment:
This invokes unsafe behavior if I do:

```rust
struct Something { value: f64 }
unsafe extern "C" fn duk_derive_foo(something: &Something) {
    println!("Definitely not valid memory: {}", something.value);
}
const duk_derive_foo_arg_len: usize = 42;

// ...
add_global_fn!(ctx, foo);
```

Reply:
This code would not type check. add_global_fn requires taking in an `unsafe extern "C" fn(*mut duk_sys::duk_context) -> i32` which `duk_derive_foo` does not satisfy.

Reply:
even if the arg_len is wrong, I don't believe it would cause any memory unsafety, as it would just refer to stack indices that do not exist, and throw an error in the duktape context.

Reply:
OK, maybe I'm mistaken!  This is going to trivially be solved by addressing the other comment chain though :)